data/src/main/java/com/microsoft/azure/kusto/data/auth/UserPromptTokenProvider.java

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.kusto.data.auth;
+
+import com.microsoft.aad.msal4j.IAccount;
+import com.microsoft.aad.msal4j.IAuthenticationResult;
+import com.microsoft.aad.msal4j.InteractiveRequestParameters;
+import com.microsoft.aad.msal4j.PublicClientApplication;
+import com.microsoft.azure.kusto.data.exceptions.DataClientException;
+import com.microsoft.azure.kusto.data.exceptions.DataServiceException;
+import org.apache.commons.lang3.StringUtils;
+import org.jetbrains.annotations.NotNull;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class UserPromptTokenProvider extends PublicAppTokenProviderBase {
+    private static final int USER_PROMPT_TIMEOUT_MS = 120 * 1000;
+    private static final URI redirectUri;
+
+    static {
+        URI tmp;
+        try {
+            tmp = new URI("http://localhost");
+        } catch (URISyntaxException e) { // This cannot happen, but is necessary to allow for a static final variable whose instantiation can throw an exception
+            tmp = null;
+        }
+        redirectUri = tmp;
+    }
+
+    private final String usernameHint;
+
+    UserPromptTokenProvider(@NotNull String clusterUrl, String authorityId) throws URISyntaxException {

```

Comment:
Must I provide a an auth id for user prompt?
Is this enforced by kcsb?

Reply:
The MSAL Java API examples always provide it, and the relevant code (com.microsoft.aad.msal4j.InteractiveRequest.java) requires one to exist - but not necessarily provided. This is because if one is not provided, it defaults to "https://login.microsoftonline.com/common/".

It is not enforced by the KCSB, so null can be provided.

Reply:
So, default = null or comment

Reply:
I think it's clear because clusterUrl is annotated as @NotNull, and authorityId isn't.

Note that Java doesn't support default method parameter values.

Reply:
I'm a subscriber to the approach of almost no comments to explain "what" (i.e. such explanations are via the code itself), and comments should instead be primarily to explain "why".

Reply:
Do you agree/comfortable with my approach, @yogilad ?

Reply:
I'll let you have this one :)